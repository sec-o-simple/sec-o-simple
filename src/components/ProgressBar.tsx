import { Fragment } from 'react/jsx-runtime'

export default function ProgressBar({
  sections,
  progress,
}: {
  sections: string[]
  progress: number
}) {
  return (
    <div className="relative mb-5 flex justify-stretch px-8 pb-1">
      {sections.map((section, i) => (
        <Fragment key={i}>
          <ProgressBarItem
            number={i + 1}
            title={section}
            isActive={progress >= i + 1}
          />
          {i < sections.length - 1 && (
            <ProgressLine progress={Math.max(0, (progress - i - 1) * 100)} />
          )}
        </Fragment>
      ))}
    </div>
  )
}

function ProgressBarItem({
  number,
  title,
  isActive,
}: {
  number: number
  title: string
  isActive?: boolean
}) {
  return (
    <div className="flex flex-col items-center justify-center gap-2">
      <div
        className={`border-default-200 bg-content1 text-neutral-foreground flex size-8 items-center justify-center rounded-full border p-4 ${
          isActive ? 'border-primary bg-primary text-primary-foreground' : ''
        }`}
      >
        {number}
      </div>
      <div
        className={`text-neutral-foreground absolute top-full text-sm ${
          isActive ? 'text-primary' : ''
        }`}
      >
        {title}
      </div>
    </div>
  )
}

function ProgressLine({ progress }: { progress: number }) {
  return (
    <div className="bg-content3 relative top-4 flex h-1 grow overflow-hidden">
      <div
        className={`bg-primary ${progress !== 100 ? 'rounded-r-full' : ''}`}
        style={{ width: `${progress}%` }}
      ></div>
    </div>
  )
}
