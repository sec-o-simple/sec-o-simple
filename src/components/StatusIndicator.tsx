export default function StatusIndicator({
  hasErrors,
  hasVisited,
}: {
  hasErrors: boolean
  hasVisited: boolean
}) {
  return (
    <div
      className={`size-2 rounded-full ${
        hasVisited
          ? hasErrors
            ? 'bg-red-400'
            : 'bg-green-500'
          : 'bg-neutral-300'
      }`}
    />
  )
}