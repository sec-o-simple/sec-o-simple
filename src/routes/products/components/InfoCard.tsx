import HSplit from '@/components/forms/HSplit'
import IconButton from '@/components/forms/IconButton'
import VSplit from '@/components/forms/VSplit'
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons'
import { HTMLProps, PropsWithChildren, ReactNode } from 'react'
import { Link } from 'react-router'

export type InfoCardProps = HTMLProps<HTMLDivElement> &
  PropsWithChildren<{
    variant?: 'boxed' | 'plain'
    title: string
    startContent?: ReactNode
    endContent?: ReactNode
    linkTo?: string
    onEdit?: () => void
    onDelete?: () => void
  }>

export default function InfoCard(props: InfoCardProps) {
  if (props.variant === 'boxed') {
    return <BoxedInfoCard {...props} />
  } else {
    const {
      title,
      startContent,
      endContent,
      linkTo,
      onEdit,
      onDelete,
      children,
      ...divProps
    } = props
    return (
      <VSplit {...divProps}>
        <HSplit className="flex items-center justify-between">
          <div className="flex items-center gap-2">
            {startContent}
            <div className="font-bold">
              {linkTo ? <Link to={linkTo}>{title}</Link> : title}
            </div>
          </div>
          <div>
            {endContent}
            {onEdit && <IconButton icon={faEdit} onPress={onEdit} />}
            {onDelete && <IconButton icon={faTrash} onPress={onDelete} />}
          </div>
        </HSplit>
        {children}
      </VSplit>
    )
  }
}

function BoxedInfoCard(props: InfoCardProps) {
  return (
    <div className="rounded-lg border bg-content1 py-2 px-4">
      <InfoCard {...props} variant="plain" />
    </div>
  )
}
