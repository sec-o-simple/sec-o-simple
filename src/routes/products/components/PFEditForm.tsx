import { Autocomplete } from '@/components/forms/Autocomplete'
import { Input } from '@/components/forms/Input'
import {
  checkReadOnly as checkTemplateReadonly,
  getPlaceholder,
} from '@/utils/template'
import { useProductTreeBranch } from '@/utils/useProductTreeBranch'
import { Button } from '@heroui/button'
import {
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
} from '@heroui/modal'
import { AutocompleteItem } from '@heroui/react'
import { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ProductFamilyChains } from '../ProductFamily'
import { TProductFamily } from '../types/tProductTreeBranch'

export type PFCreateEditFormProps = {
  pf?: TProductFamily
  onSave?: (updatedPf: TProductFamily) => void
  families?: TProductFamily[]
}

export function PFCreateEditForm({ pf, onSave }: PFCreateEditFormProps) {
  const { t } = useTranslation()
  const { families } = useProductTreeBranch()
  const { name: pfName, isReadonly } = pf
    ? { name: families.find((f) => f.id === pf.id)?.name, isReadonly: false }
    : { name: '', isReadonly: false }
  const [name, setName] = useState(pf?.name ?? '')
  const [parent, setParent] = useState(pf?.parent ?? null)

  return (
    <ModalContent>
      {(onClose) => (
        <>
          <ModalHeader>
            {pf?.id
              ? t('modal.edit', {
                  label: t('product_family.label'),
                })
              : t('modal.create', {
                  label: t('product_family.label'),
                })}
          </ModalHeader>
          <ModalBody>
            <Input
              label={t('product_family.name')}
              autoFocus
              value={isReadonly ? (pfName ?? '') : name}
              onValueChange={setName}
              isDisabled={
                pf ? checkTemplateReadonly(pf, 'name') || isReadonly : false
              }
              placeholder={pf ? getPlaceholder(pf, 'name') : undefined}
            />
            <Autocomplete
              labelPlacement="outside"
              label={t('product_family.parent')}
              selectedKey={parent?.id || ''}
              variant="bordered"
              inputProps={{
                classNames: {
                  inputWrapper: 'border-1 shadow-none',
                },
              }}
              onSelectionChange={(key) => {
                if (key === null) {
                  setParent(null)
                  return
                }

                const parentPF = families.find((f) => f.id === key)
                if (parentPF) {
                  setParent(parentPF)
                }
              }}
            >
              {families
                .filter((item) => item.id !== pf?.id)
                .map((item) => (
                  <AutocompleteItem
                    key={item.id.toString()}
                    textValue={item.name}
                  >
                    <ProductFamilyChains item={item} />
                  </AutocompleteItem>
                ))}
            </Autocomplete>
          </ModalBody>
          <ModalFooter>
            <Button onPress={onClose} variant="light">
              {t('common.cancel')}
            </Button>
            <Button
              color="primary"
              onPress={() => {
                onSave?.({
                  ...pf,
                  name,
                  parent,
                } as TProductFamily)
                onClose()
              }}
            >
              {t('common.save')}
            </Button>
          </ModalFooter>
        </>
      )}
    </ModalContent>
  )
}
