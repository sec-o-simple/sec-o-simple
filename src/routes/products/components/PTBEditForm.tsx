import {
  TProductTreeBranch,
  TProductTreeBranchProductType,
  productTreeBranchProductTypes,
} from '../types/tProductTreeBranch'
import { Button } from '@heroui/button'
import { Input, Textarea } from '@/components/forms/Input'
import { useState } from 'react'
import {
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
} from '@heroui/modal'
import { checkReadOnly, getPlaceholder } from '@/utils/template'
import Select from '@/components/forms/Select'
import { SelectItem } from '@heroui/select'
import useDocumentStore from '@/utils/useDocumentStore'

export type PTBEditFormProps = {
  ptb?: TProductTreeBranch
  onSave?: (updatedPtb: TProductTreeBranch) => void
}

export function getCategoryLabel(category: string): string {
  switch (category) {
    case 'vendor':
      return 'Vendor'
    case 'product_name':
      return 'Product'
    case 'product_version':
      return 'Product Version'
    default:
      return ''
  }
}

export function PTBEditForm({ ptb, onSave }: PTBEditFormProps) {
  const [name, setName] = useState(ptb?.name ?? '')
  const [description, setDescription] = useState(ptb?.description ?? '')
  const [type, setType] = useState(ptb?.type ?? 'Software')
  const sosDocumentType = useDocumentStore((state) => state.sosDocumentType)

  const categoryLabel = getCategoryLabel(ptb?.category ?? '')

  return (
    <ModalContent>
      {(onClose) => (
        <>
          <ModalHeader>Edit {categoryLabel}</ModalHeader>
          <ModalBody>
            <Input
              label="Name"
              autoFocus
              value={name}
              onValueChange={setName}
              isDisabled={!ptb || checkReadOnly(ptb, 'name')}
              placeholder={ptb ? getPlaceholder(ptb, 'name') : undefined}
            />
            <Textarea
              label="Description"
              value={description}
              onValueChange={setDescription}
              isDisabled={!ptb || checkReadOnly(ptb, 'description')}
              placeholder={ptb ? getPlaceholder(ptb, 'description') : undefined}
            />
            {ptb?.category === 'product_name' && (
              <Select
                label="Type"
                selectedKeys={[type ?? 'Software']}
                onChange={(e) => {
                  if (!e.target.value) {
                    return
                  }
                  setType(e.target.value as TProductTreeBranchProductType)
                }}
                isDisabled={!ptb || checkReadOnly(ptb, 'type')}
                placeholder={ptb ? getPlaceholder(ptb, 'type') : undefined}
              >
                {productTreeBranchProductTypes
                  .filter((type) => sosDocumentType.includes(type))
                  .map((type) => (
                    <SelectItem key={type}>{type}</SelectItem>
                  ))}
              </Select>
            )}
          </ModalBody>
          <ModalFooter>
            <Button onPress={onClose}>Cancel</Button>
            <Button
              color="primary"
              onPress={() => {
                if (ptb) {
                  onSave?.({ ...ptb, name, description, type })
                }
                onClose()
              }}
            >
              Save
            </Button>
          </ModalFooter>
        </>
      )}
    </ModalContent>
  )
}
