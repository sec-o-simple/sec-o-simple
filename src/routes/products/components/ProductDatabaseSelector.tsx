import { Input } from '@/components/forms/Input'
import { parseProductTree } from '@/utils/csafImport/parseProductTree'
import { parseRelationships } from '@/utils/csafImport/parseRelationships'
import { useConfigStore } from '@/utils/useConfigStore'
import {
  Vendor as DatabaseVendor,
  Product,
  TProductDatabaseCSAFProducttree,
  useDatabaseClient,
} from '@/utils/useDatabaseClient'
import useDocumentStore from '@/utils/useDocumentStore'
import { faSearch } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from '@heroui/button'
import {
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
} from '@heroui/modal'
import { Accordion, AccordionItem, Alert, Checkbox } from '@heroui/react'
import { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link } from 'react-router'
import { TProductTreeBranch } from '../types/tProductTreeBranch'

interface Props {
  isOpen: boolean
  onClose: () => void
}

type Vendor = DatabaseVendor & {
  products: Product[]
}

export default function ProductDatabaseSelector({ isOpen, onClose }: Props) {
  const client = useDatabaseClient()
  const config = useConfigStore((state) => state.config)
  const products = Object.values(useDocumentStore((state) => state.products))
  const {
    updateProducts,
    families,
    updateFamilies,
    relationships,
    updateRelationships,
  } = useDocumentStore()

  const { t } = useTranslation()
  const [selectedProducts, setSelectedProducts] = useState<string[]>([])
  const [vendors, setVendors] = useState<Vendor[]>([])
  const [dbProducts, setDBProducts] = useState<Product[]>([])

  useEffect(() => {
    async function fetchVendors() {
      const [dbVendors, products] = await Promise.all([
        client.fetchVendors(),
        client.fetchProducts(),
      ])

      setDBProducts(products)

      setVendors(
        dbVendors
          .filter((v) => products.map((p) => p.vendor_id).includes(v.id))
          .map((vendor) => {
            return {
              ...vendor,
              products: products.filter(
                (product) => product.vendor_id === vendor.id,
              ),
            }
          }),
      )
    }
    fetchVendors()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const [searchQuery, setSearchQuery] = useState('')
  const filteredVendors = vendors.filter(
    (vendor) =>
      vendor.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
      vendor.products.some((product) =>
        product.name.toLowerCase().includes(searchQuery.toLowerCase()),
      ),
  )

  const [submitting, setSubmitting] = useState(false)

  const handleAddProducts = async () => {
    if (selectedProducts.length === 0) return
    setSubmitting(true)
    try {
      // Attempt to fetch CSAF document
      let csafDocument: TProductDatabaseCSAFProducttree | undefined
      try {
        csafDocument = await client.fetchCSAFProducts(selectedProducts)
      } catch (e) {
        console.error('Error fetching CSAF document:', e)
      }

      const { products: importedPTB, families: importedFamilies } =
        parseProductTree({
          product_tree: csafDocument?.product_tree,
        })

      // Filter out existing products to avoid ID conflicts, and avoid duplicate vendors too
      // We identify vendors by their names, because Vendor-IDs are not within the CSAF standard
      const mergedPTB = [...products]
      importedPTB.forEach((importedVendor) => {
        const existingVendorIndex = mergedPTB.findIndex(
          (v) => v.name === importedVendor.name,
        )

        // Enrich products with type from database
        const importedProducts: TProductTreeBranch[] =
          importedVendor.subBranches.map((p) => ({
            ...p,
            type:
              dbProducts.find((dbP) => dbP.id === p.id)?.type === 'hardware'
                ? 'Hardware'
                : 'Software',
          }))

        if (existingVendorIndex >= 0) {
          // Vendor exists, merge products
          const existingVendor = mergedPTB[existingVendorIndex]
          const mergedSubBranches = [
            ...existingVendor.subBranches.filter((existingProduct) => {
              return !importedProducts.find(
                (impProduct) => impProduct.id === existingProduct.id,
              )
            }),
            ...importedProducts,
          ]
          mergedPTB[existingVendorIndex] = {
            ...existingVendor,
            subBranches: mergedSubBranches,
          }
        } else {
          // New vendor, add directly
          mergedPTB.push({
            ...importedVendor,
            subBranches: importedProducts,
          })
        }
      })
      updateProducts(mergedPTB)

      // Merge relationships, avoiding duplicates
      if (csafDocument?.product_tree?.relationships) {
        const newRelationships = parseRelationships(
          csafDocument?.product_tree?.relationships,
          mergedPTB,
        )
        const mergedRelationships = [
          ...relationships.filter((rel) => {
            return !newRelationships.find(
              (newRel) =>
                newRel.productId1 === rel.productId1 &&
                newRel.productId2 === rel.productId2 &&
                newRel.category === rel.category,
            )
          }),
          ...newRelationships,
        ]
        updateRelationships(mergedRelationships)
      }

      // Filter out existing families to avoid ID conflicts
      const mergedFamilies = [
        ...families.filter((family) => {
          return !importedFamilies.find(
            (impFamily) => impFamily.id === family.id,
          )
        }),
        ...importedFamilies,
      ]
      updateFamilies(mergedFamilies)
    } catch (error) {
      console.error('Failed to add products:', error)
    } finally {
      setSubmitting(false)
      setSelectedProducts([])
      setSearchQuery('')
      onClose()
    }
  }

  return (
    <Modal isOpen={isOpen} onClose={onClose} size="xl" scrollBehavior="inside">
      <ModalContent>
        {(onClose) => (
          <>
            <ModalHeader className="flex flex-col gap-1">
              {t('products.import.title')}
            </ModalHeader>
            <ModalBody>
              <p>{t('products.import.description')}</p>
              <Alert className="mt-2" color="default">
                <p>
                  {t('products.import.warning')}{' '}
                  <Link
                    to={config?.productDatabase?.url || '#'}
                    className="underline"
                    target="_blank"
                  >
                    {t('products.import.database')}
                  </Link>
                </p>
              </Alert>
              <div className="py-4">
                <Input
                  placeholder={t('products.import.searchPlaceholder')}
                  value={searchQuery}
                  onChange={(e) => setSearchQuery(e.target.value)}
                  className="mb-2"
                  startContent={
                    <FontAwesomeIcon
                      icon={faSearch}
                      className="text-slate-500"
                    />
                  }
                />

                {vendors.length === 0 && (
                  <Alert color="warning">
                    {t('products.import.noVendors')}
                  </Alert>
                )}

                <Accordion variant="shadow">
                  {filteredVendors.map((vendor) => {
                    const vendorProducts = vendor.products.map(
                      (product) => product.id,
                    )
                    const selectedProductCount = selectedProducts.filter((id) =>
                      vendorProducts.includes(id),
                    ).length
                    const selectedAllProducts =
                      selectedProductCount === vendor.products.length

                    return (
                      <AccordionItem
                        key={vendor.id}
                        title={vendor.name}
                        subtitle={
                          selectedProductCount
                            ? selectedAllProducts
                              ? t('products.import.selectedAll')
                              : t('products.import.selected', {
                                  count: selectedProductCount,
                                })
                            : ''
                        }
                        startContent={
                          <Checkbox
                            isSelected={selectedAllProducts}
                            onChange={() => {
                              setSelectedProducts((prev) => {
                                if (selectedAllProducts) {
                                  return prev.filter(
                                    (id) => !vendorProducts.includes(id),
                                  )
                                } else {
                                  return [...prev, ...vendorProducts]
                                }
                              })
                            }}
                          />
                        }
                      >
                        <div className="ml-4 flex flex-col gap-2 pb-4">
                          {vendor.products.map((product) => (
                            <Checkbox
                              key={product.id}
                              isSelected={selectedProducts.includes(product.id)}
                              onChange={(e) => {
                                const isChecked = e.target.checked
                                setSelectedProducts((prev) =>
                                  isChecked
                                    ? [...prev, product.id]
                                    : prev.filter((id) => id !== product.id),
                                )
                              }}
                            >
                              {product.name}
                            </Checkbox>
                          ))}
                        </div>
                      </AccordionItem>
                    )
                  })}
                </Accordion>
              </div>
            </ModalBody>
            <ModalFooter>
              <Button variant="light" onPress={onClose}>
                {t('common.cancel')}
              </Button>
              <Button
                color="primary"
                isLoading={submitting}
                onPress={handleAddProducts}
                isDisabled={selectedProducts.length === 0}
              >
                {t('products.import.add', { count: selectedProducts.length })}
              </Button>
            </ModalFooter>
          </>
        )}
      </ModalContent>
    </Modal>
  )
}
