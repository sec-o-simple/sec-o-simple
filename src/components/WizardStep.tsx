import { Button } from '@heroui/button'
import { PropsWithChildren } from 'react'
import { useNavigate } from 'react-router'
import ProgressBar from './ProgressBar'
import { useTranslation } from 'react-i18next'

export type WizardStepProps = PropsWithChildren<{
  title?: string
  progress?: number
  onBack?: string | (() => void)
  onContinue?: string | (() => void)
  noContentWrapper?: boolean
}>

export default function WizardStep({
  title,
  progress,
  onBack,
  onContinue,
  children,
  noContentWrapper,
}: WizardStepProps) {
  const navigate = useNavigate()
  const { t } = useTranslation()

  return (
    <div className="flex max-w-5xl flex-col gap-4 p-8">
      <ProgressBar
        sections={[
          t('nav.document'),
          t('nav.products'),
          t('nav.vulnerabilities'),
        ]}
        progress={progress ?? 1}
      />
      {(noContentWrapper && <>{children}</>) || (
        <div className="flex flex-col gap-4 rounded-lg bg-content1 p-8 shadow">
          {title && <div className="mb-2 text-xl font-semibold">{title}</div>}
          {children}
        </div>
      )}
      <div className="flex justify-between">
        <div>
          {onBack !== undefined && (
            <Button
              onPress={() =>
                typeof onBack === 'string' ? navigate(onBack) : onBack?.()
              }
              variant="bordered"
              className="border-1 bg-content1"
            >
              {t('common.back')}
            </Button>
          )}
        </div>
        {onContinue !== undefined && (
          <Button
            onPress={() =>
              typeof onContinue === 'string'
                ? navigate(onContinue)
                : onContinue?.()
            }
            color="primary"
          >
            {t('common.next')}
          </Button>
        )}
      </div>
    </div>
  )
}
