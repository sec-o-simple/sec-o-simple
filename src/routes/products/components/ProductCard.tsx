import { Chip } from '@heroui/chip'
import { TProductTreeBranch } from '../types/tProductTreeBranch'
import InfoCard, { InfoCardProps } from './InfoCard'
import TagList from './TagList'

export type ProductCardProps = Partial<InfoCardProps> & {
  product: TProductTreeBranch
}

export default function ProductCard({ product, ...props }: ProductCardProps) {
  return (
    <InfoCard
      variant="boxed"
      title={product.name}
      linkTo={`product/${product.id}`}
      startContent={
        <Chip color="primary" variant="flat" radius="md" size="lg">
          {product.type}
        </Chip>
      }
      {...props}
    >
      <TagList tags={product.subBranches.map((version) => version.name)} />
    </InfoCard>
  )
}
