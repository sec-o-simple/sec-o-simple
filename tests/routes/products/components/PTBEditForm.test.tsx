import { render } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'

// Mock @heroui/react to include Autocomplete
vi.mock('@heroui/react', async (importOriginal) => {
  const actual = (await importOriginal()) as any
  return {
    ...actual,
    Autocomplete: ({ children, ...props }: any) => (
      <div data-testid="autocomplete" {...props}>
        {children}
      </div>
    ),
    AutocompleteItem: ({ children, ...props }: any) => (
      <div data-testid="autocomplete-item" {...props}>
        {children}
      </div>
    ),
  }
})

import { PTBCreateEditForm } from '../../../../src/routes/products/components/PTBEditForm'
import type { TProductTreeBranch } from '../../../../src/routes/products/types/tProductTreeBranch'

// Mock dependencies
vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string, options?: any) => {
      if (options?.label) return `${key} ${options.label}`
      return key
    },
  }),
}))

vi.mock('../../../../src/utils/useDocumentType', () => ({
  default: () => ({
    hasHardware: true,
    hasSoftware: true,
  }),
}))

vi.mock('../../../../src/utils/template', () => ({
  checkReadOnly: vi.fn(() => false),
  getPlaceholder: vi.fn(() => 'Placeholder text'),
}))

// Mock useProductTreeBranch hook - override global mock
vi.unmock('../../../../src/utils/useProductTreeBranch')
vi.mock('../../../../src/utils/useProductTreeBranch', () => ({
  useProductTreeBranch: vi.fn(() => ({
    getPTBName: vi.fn((branch) => ({
      name: branch.name,
      isReadonly: false,
    })),
    families: [
      { id: 'family1', name: 'Test Family 1' },
      { id: 'family2', name: 'Test Family 2' },
    ],
    getRelationshipFullProductName: vi.fn(() => 'Mock Product Name'),
    getFullProductName: vi.fn(() => 'Mock Full Product Name'),
  })),
}))

// Mock HeroUI components
vi.mock('@heroui/modal', () => ({
  ModalContent: ({ children }: { children: any }) => (
    <div data-testid="modal-content">
      {typeof children === 'function' ? children(() => {}) : children}
    </div>
  ),
  ModalHeader: ({ children }: { children: any }) => (
    <div data-testid="modal-header">{children}</div>
  ),
  ModalBody: ({ children }: { children: any }) => (
    <div data-testid="modal-body">{children}</div>
  ),
  ModalFooter: ({ children }: { children: any }) => (
    <div data-testid="modal-footer">{children}</div>
  ),
}))

vi.mock('@heroui/button', () => ({
  Button: ({ children, ...props }: any) => (
    <button {...props} data-testid="button">
      {children}
    </button>
  ),
}))

vi.mock('@heroui/select', () => ({
  SelectItem: ({ children, ...props }: any) => (
    <option {...props} data-testid="select-item">
      {children}
    </option>
  ),
}))

// Mock custom form components
vi.mock('../../../../src/components/forms/Input', () => ({
  Input: (props: any) => <input data-testid="input" {...props} />,
  Textarea: (props: any) => <textarea data-testid="textarea" {...props} />,
}))

vi.mock('../../../../src/components/forms/Select', () => ({
  default: ({ children, ...props }: any) => (
    <select data-testid="select" {...props}>
      {children}
    </select>
  ),
}))

describe('PTBCreateEditForm', () => {
  const mockPTB: TProductTreeBranch = {
    id: 'test-ptb-1',
    category: 'product_name',
    name: 'Test Product',
    description: 'Test product description',
    type: 'Software',
    subBranches: [],
  }

  it('should render create form and match snapshot', () => {
    const { container } = render(
      <PTBCreateEditForm category="product_name" onSave={vi.fn()} />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render edit form with existing PTB and match snapshot', () => {
    const { container } = render(
      <PTBCreateEditForm
        ptb={mockPTB}
        category="product_name"
        onSave={vi.fn()}
      />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render vendor form (without description and type fields) and match snapshot', () => {
    const vendorPTB: TProductTreeBranch = {
      id: 'test-vendor-1',
      category: 'vendor',
      name: 'Test Vendor',
      description: 'Test vendor description',
      subBranches: [],
    }

    const { container } = render(
      <PTBCreateEditForm ptb={vendorPTB} category="vendor" onSave={vi.fn()} />,
    )

    expect(container).toMatchSnapshot()
  })
})
