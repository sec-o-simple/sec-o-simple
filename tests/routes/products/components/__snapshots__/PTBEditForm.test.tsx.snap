// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PTBCreateEditForm > should render create form and match snapshot 1`] = `
<div>
  <div
    data-testid="modal-content"
  >
    <div
      data-testid="modal-header"
    >
      modal.create product_name.label
    </div>
    <div
      data-testid="modal-body"
    >
      <input
        data-testid="input"
        label="product_name.name"
        value=""
      />
      <textarea
        data-testid="textarea"
        label="product_name.description"
      />
      <select
        data-testid="select"
        label="product_name.type"
        selectedkeys="Software"
      >
        <option
          data-testid="select-item"
        >
          Software
        </option>
        <option
          data-testid="select-item"
        >
          Hardware
        </option>
      </select>
    </div>
    <div
      data-testid="modal-footer"
    >
      <button
        data-testid="button"
        variant="light"
      >
        common.cancel
      </button>
      <button
        color="primary"
        data-testid="button"
      >
        common.save
      </button>
    </div>
  </div>
</div>
`;

exports[`PTBCreateEditForm > should render edit form with existing PTB and match snapshot 1`] = `
<div>
  <div
    data-testid="modal-content"
  >
    <div
      data-testid="modal-header"
    >
      modal.edit product_name.label
    </div>
    <div
      data-testid="modal-body"
    >
      <input
        data-testid="input"
        label="product_name.name"
        placeholder="Placeholder text"
        value="Test Product"
      />
      <textarea
        data-testid="textarea"
        label="product_name.description"
        placeholder="Placeholder text"
      >
        Test product description
      </textarea>
      <select
        data-testid="select"
        label="product_name.type"
        placeholder="Placeholder text"
        selectedkeys="Software"
      >
        <option
          data-testid="select-item"
        >
          Software
        </option>
        <option
          data-testid="select-item"
        >
          Hardware
        </option>
      </select>
    </div>
    <div
      data-testid="modal-footer"
    >
      <button
        data-testid="button"
        variant="light"
      >
        common.cancel
      </button>
      <button
        color="primary"
        data-testid="button"
      >
        common.save
      </button>
    </div>
  </div>
</div>
`;

exports[`PTBCreateEditForm > should render vendor form (without description and type fields) and match snapshot 1`] = `
<div>
  <div
    data-testid="modal-content"
  >
    <div
      data-testid="modal-header"
    >
      modal.edit vendor.label
    </div>
    <div
      data-testid="modal-body"
    >
      <input
        data-testid="input"
        label="vendor.name"
        placeholder="Placeholder text"
        value="Test Vendor"
      />
    </div>
    <div
      data-testid="modal-footer"
    >
      <button
        data-testid="button"
        variant="light"
      >
        common.cancel
      </button>
      <button
        color="primary"
        data-testid="button"
      >
        common.save
      </button>
    </div>
  </div>
</div>
`;
